### PR TITLE
nghttp2: optionally enable python3 bindings

### PR DESCRIPTION
**Pull request by fwiesel**
This optionally enables the python 3 bindings of the nghttp2 library.
It should still work with python2, when with-docs are requested. 

The patch is required to detect the linker flags properly under Mac OS X and python 3.4.
I'll try to upstream it, as soon as I find out, where in the automake chain that actually gets generated.

### CODE REVIEW

File: Library/Formula/nghttp2.rb [hunk 1]
```diff
@@ -20,8 +20,10 @@ class Nghttp2 < Formula
 
   option "with-examples", "Compile and install example programs"
   option "without-docs", "Don't build man pages"
+  option "with-python3", "This is required for enabling the python bindings"
 
-  depends_on :python => :build if MacOS.version <= :snow_leopard && build.with?("docs")
+  depends_on :python => :build if MacOS.version <= :snow_leopard && build.with?("docs") && build.without?("python3")
+  depends_on :python3 => :optional
```
Comment (mikemcquaid) [inline Library/Formula/nghttp2.rb]: Is this not also a build-time requirement?

File: Library/Formula/nghttp2.rb [hunk 2]
```diff
@@ -87,35 +89,66 @@ class Nghttp2 < Formula
...
+   end
+  end
+
+  def python
+   @python ||= begin
+     "python#{python_version.join('.')}"
+   end
+  end
+
+  patch :p1, :DATA
```
Comment (mikemcquaid) [inline Library/Formula/nghttp2.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

**Comment by mikemcquaid** [comment]
I'd like to get @tdsmith's Python views here. My personal view is that it's OK to build Python 3 bindings but it's better if the actual wrapper scripts are using Python 2.

File: Library/Formula/nghttp2.rb [hunk 3]
```diff
@@ -87,35 +89,66 @@ class Nghttp2 < Formula
     sha256 "3e15b416c9a2039c1a51208b2cd3bb4ffd796cd19e601b1d2657afcb77c3dc90"
   end
 
+  resource "Cython" do
+    url "https://pypi.python.org/packages/source/C/Cython/cython-0.22.tar.gz"
+    sha256 "14307e7a69af9a0d0e0024d446af7e51cc0e3e4d0dfb10d36ba837e5e5844015"
+  end if build.with?("python3")
```
Comment (DomT4) [inline Library/Formula/nghttp2.rb]: You don't need to protectively wrap resources. They don't get activated unless you call them.

File: Library/Formula/nghttp2.rb [hunk 4]
```diff
@@ -87,35 +89,66 @@ class Nghttp2 < Formula
...
     ENV.cxx11
 
     if build.with? "docs"
-      ENV.prepend_create_path "PYTHONPATH", buildpath+"sphinx/lib/python2.7/site-packages"
+      ENV.prepend_create_path "PYTHONPATH", buildpath+"sphinx/lib/#{python}/site-packages"
       resources.each do |r|
         r.stage do
-          system "python", *Language::Python.setup_install_args(buildpath/"sphinx")
-        end
+          system python, *Language::Python.setup_install_args(buildpath/'sphinx')
```
Comment (DomT4) [inline Library/Formula/nghttp2.rb]: Please don't change double quotes into single quotes.

File: Library/Formula/nghttp2.rb [hunk 5]
```diff
@@ -87,35 +89,66 @@ class Nghttp2 < Formula
...
+        end unless r.name == 'Cython'
```
Comment (DomT4) [inline Library/Formula/nghttp2.rb]: Likewise, please change single quotes into double quotes.

**Comment by tdsmith** [comment]
Optionally building bindings is great! Some notes:

* The docs should always be built with (system) python2
* if cython is only needed at build time, it shouldn't be installed to libexec
* if cython is needed at runtime, you need to write a .pth file to `lib/python#{Language::Python.major_minor_version "python3"}/site-packages` that adds the cython site-packages to sys.path
* python3-config --ldflags does not give the best ldflags for building python modules (use `-undefined dynamic_lookup` instead of `-lpython` or `-framework Python`)
* assuming python2 is 2.7 is safe; assuming python3 is 3.4 is not (3.5 will come out this year)

File: Library/Formula/nghttp2.rb [hunk 6]
```diff
@@ -87,35 +89,66 @@ class Nghttp2 < Formula
...
       --enable-app
       --with-boost=#{Formula["boost"].opt_prefix}
       --enable-asio-lib
-      --disable-python-bindings
     ]
 
     args << "--enable-examples" if build.with? "examples"
     args << "--with-spdylay" if build.with? "spdylay"
+    if build.with?('python3')
+      args << "--enable-python-bindings" << "PYTHON=#{Formula['python3'].bin}/#{python}" << "CYTHON=#{libexec}/vendor/bin/cython"
```
Comment (tdsmith) [inline Library/Formula/nghttp2.rb]: just use `"PYTHON=python3"`; the `:python3` dependency uses the python3 from PATH, which may not be a brewed python

**Comment by tdsmith** [comment]
Are the bindings python3-only?

File: Library/Formula/nghttp2.rb (hunk 1 above)
Comment (fwiesel) [inline Library/Formula/nghttp2.rb]: Well, when installing the python bindings, it is a runtime requirement.

**Comment by fwiesel** [comment]
As I understand their [python api](https://nghttp2.org/documentation/python-apiref.html#http-2-servers), they are python3 only, due to the asyncio requirement.

**Comment by fwiesel** [comment]
I hope, that fits better. 

Instead of patching, I override the extra flags with `-undefined dynamic_lookup`.

I thought using two different python versions with overlapping PYTHONPATH is bound to bring some trouble, but on the other hand, it is a fairly limited set of libraries.

**Comment by DomT4** [comment]
This looks okay now but will let Tim :+1: or :-1:.

**Comment by tdsmith** [comment]
> I thought using two different python versions with overlapping PYTHONPATH is bound to bring some trouble

Hmmm, that's a good point; we could work around it but this looks like it's probably okay.

**Comment by tdsmith** [comment]
Thanks for all the changes; this looks great. Pulled in 3ac0294! Thanks for the contribution. :tada:

**Comment by RandomDSdevel** [comment]
>> …
>>
>>I thought using two different python versions with overlapping PYTHONPATH is bound to bring some trouble[.]  
>
>Hmmm, that's a good point; we could work around it but this looks like it's probably okay.

Speaking of which, here's what you get if you try to install nghttp2 via Homebrew on a system that has both Python 2.x and Python 3.x installed:  

```
Bryces-HD-1-TB:~ zadmin$ brew install -vd --build-from-source nghttp2 --with-examples --with-python3
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/nghttp2.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/xz.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/pkg-config.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/cunit.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/autoconf.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/automake.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libtool.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libev.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/openssl.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/makedepend.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libevent.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/jansson.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/boost.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/spdylay.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libxml2.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/doxygen.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/cmake.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Library/Formula/nghttp2.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libev.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/openssl.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libevent.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/jansson.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/boost.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/spdylay.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libxml2.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/xz.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/pkg-config.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/cunit.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/autoconf.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/automake.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libtool.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/makedepend.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/python3.rb
==> Downloading https://github.com/tatsuhiro-t/nghttp2/releases/download/v1.3.0/nghttp2-1.3.0.tar.xz
Already downloaded: /Library/Caches/Homebrew/nghttp2-1.3.0.tar.xz
==> Verifying nghttp2-1.3.0.tar.xz checksum
==> Downloading https://pypi.python.org/packages/source/S/Sphinx/Sphinx-1.3.1.tar.gz
Already downloaded: /Library/Caches/Homebrew/nghttp2--sphinx-1.3.1.tar.gz
==> Verifying nghttp2--sphinx-1.3.1.tar.gz checksum
tar xf /Library/Caches/Homebrew/nghttp2--sphinx-1.3.1.tar.gz
==> python -c import setuptools... --no-user-cfg install --prefix=/private/tmp/nghttp220150901-28972-1jc3fyk/nghttp2-1.3.0/sphinx --single-version-externally-managed --record=installed.txt
Your PYTHONPATH points to a site-packages dir for Python 3.x but you are running Python 2.x!
     PYTHONPATH is currently: "/private/tmp/nghttp220150901-28972-1jc3fyk/nghttp2-1.3.0/sphinx/lib/python2.7/site-packages:/usr/local/lib/python3.4/site-packages"
     You should `unset PYTHONPATH` to fix this.
/usr/local/Library/Homebrew/debrew.rb:11:in `raise'
BuildError: Failed executing: python -c import\ setuptools,\ tokenize
__file__\ =\ 'setup.py'
exec(compile(getattr(tokenize,\ 'open',\ open)(__file__).read()
\ \ .replace('\r\n',\ '\n'),\ __file__,\ 'exec'))
 --no-user-cfg install --prefix=/private/tmp/nghttp220150901-28972-1jc3fyk/nghttp2-1.3.0/sphinx --single-version-externally-managed --record=installed.txt
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 1
==> Formula
Path: /usr/local/Library/Formula/nghttp2.rb
==> Configuration
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: 6d65e3abbc6db8b85a394e796287868d23bb311c
Last commit: 48 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: dual-core 64-bit core2
OS X: 10.10.5-x86_64
Xcode: 6.4
CLT: 6.4.0.0.1.1435007323
Clang: 6.1 build 602
X11: 2.7.7 => /opt/X11
System Ruby: 2.0.0-p481
Perl: /usr/local/bin/perl => /usr/local/Cellar/perl/5.22.0/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/local/bin/ruby => /usr/local/Cellar/ruby/2.2.3/bin/ruby
Java: 1.8.0_60
==> ENV
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
MAKEFLAGS: -j2
CMAKE_PREFIX_PATH: /usr/local/opt/openssl:/usr/local/opt/libxml2:/usr/local
CMAKE_INCLUDE_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
CMAKE_FRAMEWORK_PATH: /usr/local/opt/python3/Frameworks
PKG_CONFIG_PATH: /usr/local/opt/xz/lib/pkgconfig:/usr/local/opt/cunit/lib/pkgconfig:/usr/local/opt/openssl/lib/pkgconfig:/usr/local/opt/libevent/lib/pkgconfig:/usr/local/opt/jansson/lib/pkgconfig:/usr/local/opt/libxml2/lib/pkgconfig:/usr/local/opt/spdylay/lib/pkgconfig:/usr/local/opt/python3/lib/pkgconfig
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Library/ENV/pkgconfig/10.10
ACLOCAL_PATH: /usr/local/opt/libxml2/share/aclocal:/usr/local/share/aclocal
PATH: /usr/local/Library/ENV/4.3:/usr/local/opt/xz/bin:/usr/local/opt/pkg-config/bin:/usr/local/opt/openssl/bin:/usr/local/opt/libevent/bin:/usr/local/opt/libxml2/bin:/usr/local/opt/spdylay/bin:/usr/local/opt/python3/bin:/usr/bin:/bin:/usr/sbin:/sbin

Error: nghttp2 1.3.0 did not build
Logs:
     /Users/zadmin/Library/Logs/Homebrew/nghttp2/01.python
```

Should we open a new issue for this, or am I just doing something wrong?

**Comment by tdsmith** [comment]
This broke because of some recent improvements to our Python handling.

**Comment by RandomDSdevel** [comment]
@tdsmith:  Huh, figures.

**Pull request closed without merging**